lib/stages/BeamBufferSort.cpp

```diff
@@ -0,0 +1,295 @@
+        nonfreq_meta-> ctime, nonfreq_meta -> stream_id.id, nonfreq_meta -> dataset_id,\
+	nonfreq_meta -> beam_number, nonfreq_meta -> ra, nonfreq_meta -> dec,\
+	nonfreq_meta -> scaling);
+}
+
+void BeamBufferSort::fill_empty_frame(uint32_t time_idx, uint32_t freq_idx, 
+    uint64_t fpga_seq_start0, timespec ctime0, uint32_t beam_number, double ra, 
+    double dec, double scaling){
+    
+    uint32_t fill_pos;
+    uint64_t fake_fpga_seq_start;
+    uint32_t fake_freq_bin;     // For metadata in the empty frame
+    timespec fake_ctime;
+
+    fill_pos = freq_idx * sub_frame_size;
+    fake_fpga_seq_start = fpga_seq_start0 + time_idx;
+    fake_ctime = ctime0;
+    add_nsec(fake_ctime, (long)(time_idx * subframe_time_nsec));
+    fake_freq_bin = freq_idx;
+    make_freq_meta((FreqIDBeamMetadata*) &sort_queue[time_idx][fill_pos], fake_freq_bin,
+        fake_fpga_seq_start, fake_ctime, 0, dset_id_t::null, beam_number, ra, dec, scaling);
+    memset(&sort_queue[time_idx][fill_pos + FreqIDBeamMeta_size], 0, samples_per_data_set);
+    queue_status[time_idx][freq_idx] = 1;
+}
+
+void BeamBufferSort::main_thread(){
+    uint32_t freq_bin;
+    int time_offset;
+    uint32_t start_pos;
+    uint64_t frame_nr;
+    uint8_t* output;
+    uint32_t dump_nframe;
+    uint32_t dump_start;
+    uint32_t dump_size;
+    uint64_t fpga_seq_start;
+    uint64_t fpga_seq_start0;
+    uint64_t frame0 = 0;
+    uint32_t beam_number;
+    double ra;
+    double dec;
+    double scaling;
+
+    timespec ctime0;
+    
+    FreqIDBeamMetadata* in_metadata = nullptr;
+    BeamMetadata* non_freq_metadata = nullptr;
+
+    frameID input_frame_id(in_buf); // Input frame id
+    std::vector<frameID> output_frame_ids; 
+    for (auto& out_buf : out_bufs){
+        output_frame_ids.push_back(frameID(out_buf));
+    }
+    
+    while (!stop_thread) {
+	// Receive a new frame
+        uint8_t* input = wait_for_full_frame(in_buf, unique_name.c_str(), input_frame_id);
+	if (input == nullptr)
+            break;
+        
+	// Get the metadata for each frame.
+        if (has_freq_bin) {
+            // Frequency ID is in the metdata.
+            in_metadata = (FreqIDBeamMetadata*)get_metadata(in_buf, input_frame_id);
+            freq_bin = in_metadata -> frequency_bin;
+            fpga_seq_start = in_metadata -> fpga_seq_start;
+        }
+        else {
+	    // Frequency ID is not in the metadata. will add it to the metadata.
+            non_freq_metadata = (BeamMetadata*)get_metadata(in_buf, input_frame_id);
+            fpga_seq_start = non_freq_metadata -> fpga_seq_start;
+            const uint32_t num_freq_per_stream = Telescope::instance().num_freq_per_stream();
+            for (uint32_t f = 0; f < num_freq_per_stream; ++f) {
+                freq_bin = Telescope::instance().to_freq_id(non_freq_metadata -> stream_id, f);
+            }
+	    // Build the full Frequency IDed metadata
+	    in_metadata = (FreqIDBeamMetadata*)malloc(FreqIDBeamMeta_size);
```

Comment:
`in_metadata` gets created for each new frame, and never gets freed.  This will create a memory leak that will eventually crash the program.   Ideally you should just create the space for this once at startup, and then free it in the destructor (if you needed it).   Since the same variable is used for both type of metadata you have to be careful to only free in the case you created it, and not in the case where you change this pointer to what is returned by `get_metadata()`